### DISCUSSION TITLE
How to find the closest point on an element in an iModel to a mouse-clicked point?

### DISCUSSION THREAD

**Discussion by MYDJH** [iTwin/itwinjs-core#6949] [Q&A]
Hello, does itwin.js provide a method to find the coordinates of the closest point on an iModel element to a mouse-clicked point in the interface?

It's somewhat similar to const curve = LineSegment3d.create(points) curve.closestPoint(point, false); where closestPoint method is used.

If you have any suggestions, I would greatly appreciate it
![snipaste_20240710_180956](https://github.com/iTwin/itwinjs-core/assets/101549988/7f8c33c2-89d1-4892-95f7-71b8c3d2b679)

**Comment by bbastings** [comment]
There isn't a method to return the closest point given an element id if that is what you are asking. There are closest point methods for various GeometryQuery sub-classes though, like CurvePrimitive.closestPoint, CurveCollection.closestPoint, and PolyfaceRangeTreeContext.searchForClosestPoint that can be used like the LineSegment3d.closestPoint method you showed. You'll need to extract/iterate the element's geometry.

**Reply by MYDJH** [reply]
Hello, I am very grateful for your advice. As you mentioned, I need to iterate over this element.I can access the GeometryStream within the element's structure, and then by using
const iterator = new GeometryStreamIterator(GeometryStream); 
 console.log(iterator);
![image](https://github.com/iTwin/itwinjs-core/assets/101549988/361919ba-ea77-49f4-853c-beedb3fb9b8c)
I've found that it doesn't contain the geometric information I was looking for.
How should I correctly iterate through it to obtain the geometric information? Do you have any suggestions?
The GeometryStream available within the element：
![image](https://github.com/iTwin/itwinjs-core/assets/101549988/ca8688fb-0b8b-4c22-95c3-658a006c982a)

**Reply by bbastings** [reply]
You'll want to get the GeometricElementProps and request that the geometry be returned in json format, then you can use the GeometryStreamIterator.

```
    const props = imodel.elements.getElementProps<GeometricElement3dProps>({ id: newId, wantGeometry: true });
    assert.isDefined(props.geom);

    const it = new GeometryStreamIterator(props.geom, props.category);
    for (const entry of it) {
      assert.equal(entry.primitive.type, "geometryQuery");
...
    }
```

**Reply by MYDJH** [reply]
Thank you very much for your help.

**Comment by kaushandutta** [comment]
Hi @MYDJH can you tell how can I get the information about the point I clicked on the imodel ?

**Reply by ben-polinsky** [reply]
Try [ElementLocateManager.doLocate](https://www.itwinjs.org/reference/core-frontend/locatingelements/elementlocatemanager/#dolocate). You can get an instance of the `ElementLocateManager` through `iModelApp.locateManager`
